dlog.8
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "DLOG" "8" "" ""
.SH NAME
dlog \- Send lines from standard input to a log file
.SH SYNOPSIS
.sp
\fBdlog [options] [logfile]\fP
.SH DESCRIPTION
.sp
The \fBdlog\fP program sends lines given as standard input to a log file,
one line at a time, optionally adding a timestamp in front of each line.
If the log file is not specified, then lines are printed back to standard
output. The latter may be useful to add timestamps in shell pipelines.
.SH USAGE
.sp
Command line options:
.INDENT 0.0
.TP
.BI \-p \ TEXT\fR,\fB \ \-\-prefix \ TEXT
Insert the given text as prefix for each logged message. If
adding timestamps is enabled, the text is inserted \fIafter\fP
the timestamp, but still before the logged text.
.TP
.BI \-i \ NUMBER\fR,\fB \ \-\-input\-fd \ NUMBER
Use file descriptor \fBNUMBER\fP to read input. By default the
standard input descriptor (number \fB0\fP) is used.
.TP
.B  \-b\fP,\fB  \-\-buffered
Buffered operation. If enabled, calls to \fIfsync(2)\fP will be
avoided. This improves performance, but may cause messages to
be lost.
.TP
.B  \-t\fP,\fB  \-\-timestamp
Prepend a timestamp to each saved line. By default
timestamps are disabled. Timestamp format is
\fBYYYY\-mm\-dd/HH:MM:SS\fP\&.
.TP
.B  \-e\fP,\fB  \-\-skip\-empty
Ignore empty input lines. An empty line is one that does not
contain any characters; a line which contains whitespace is
\fBnot\fP considered empty.
.TP
.B  \-h\fP,\fB  \-\-help
Show a summary of available options.
.UNINDENT
.sp
Albeit it can be used stand\-alone, most of the time you will be running
\fBdlog\fP under a process control tool like \fIdmon(8)\fP or \fIsupervise(8)\fP\&.
.SH ENVIRONMENT
.sp
Additional options will be picked from the \fBDLOG_OPTIONS\fP environment
variable, if defined. Any command line option can be specified this way.
Arguments read from the environment variable will be prepended to the ones
given in the command line, so they may still be overriden.
.SH SEE ALSO
.sp
\fIdmon(8)\fP, \fIdslog(8)\fP, \fIrotlog(8)\fP, \fImultilog(8)\fP, \fIsupervise(8)\fP
.sp
 <http://cr.yp.to/daemontools.html> 
.SH AUTHOR
Adrian Perez <aperez@igalia.com>
.\" Generated by docutils manpage writer.
.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [v0.6.0] - 2024-12-10
### Added
- New `denv` utility, inspired by daemontools' `envdir`, which may be
  used to manipulate the environment before executing a chained child
  process.

### Fixed
- Retry system calls that may return with `errno` set to `EINTR` when
  interrupted by signals to indicate they can be restarted from user space.
- Remove leftover, useless `$E` format specifier when reporting errors
  during daemonization.

### Changed
- The project is now built in C23 mode by default. In practice `-std=c2x`
  gets used to cover systems which may have slightly older compilers.

## [v0.5.1] - 2021-02-23
### Added
- The `dlog`, `dslog`, and `drlog` tools now log empty input lines. The
  `--skip-empty`/`-e` command line option has been added to disable
  logging empty lines.
- The `dmon` tool has gained a `--max-respawns`/`-m` command line option which
  can be used to specify how many times to respawn monitored processes
  before exiting. (Patch by Matt Schulte <<mschulte@wyze.com>>.)
- New `libsetunbuf.so` helper which can be used to disable buffering of
  the standard output stream on arbitrary programs via `LD_PRELOAD`.
  (Patch by Matt Schulte <<mschulte@wyze.com>>.)

### Fixed
- The exit status of monitored processes is now correctly propagated as
  exit code of `dmon` itself.
- The `dlog`, `dslog`, and `drlog` tools will no longer exit unexpectedly
  when they receive an empty input line. (Patch by Matt Schulte
  <<mschulte@wyze.com>>.)

## [v0.5.0] - 2020-08-27
### Added
- Support listing applets compiled into a multicall `dmon` binary when the
  `DMON_LIST_MULTICALL_APPLETS` environment variable is set and non-zero.

### Changed
- GNU Make is not required anymore, the included `Makefile` now works with
  the BSD variant as well, and probably others.

## [v0.4.5] - 2018-08-05
### Fixed
- Make it possible to build `dmon` with newer GCC versions.

## [v0.4.4] - 2016-10-30
### Fixed
- Correctly forward signals to the log process when using `-S`/`--log-sigs`.

## [v0.4.3] - 2016-10-13
### Added
- Allow setting the work directory with `-W`/`--work-dir` in the command line
  (or `work-dir` in a configuration file).

## [v0.4.2] - 2012-01-07
### Added
- Manual pages now also include the long command line options.
- `dlog` can now be instructed to prefix every log message with a given string,
  using the `-p`/`--prefix` command line option.
- `dlog`, `drlog`, and `dslog` have gained the `-i`/`--input-fd` command line
  option, which allows reading log messages from file descriptors different from
  the standard input.

### Fixed
- `dlog` and `drlog` now handle the `INT`, `TERM`, and `HUP` signals gracefully.

[Unreleased]: https://github.com/aperezdc/dmon/compare/v0.6.0...HEAD
[v0.6.0]: https://github.com/aperezdc/dmon/compare/v0.5.1...v0.6.0
[v0.5.1]: https://github.com/aperezdc/dmon/compare/v0.5.0...v0.5.1
[v0.5.0]: https://github.com/aperezdc/dmon/compare/v0.4.5...v0.5.0
[v0.4.5]: https://github.com/aperezdc/dmon/compare/v0.4.4...v0.4.5
[v0.4.4]: https://github.com/aperezdc/dmon/compare/v0.4.3...v0.4.4
[v0.4.3]: https://github.com/aperezdc/dmon/compare/v0.4.2...v0.4.3
[v0.4.2]: https://github.com/aperezdc/dmon/compare/v0.4.1...v0.4.2


dlog.c
/*
 * dlog.c
 * Copyright (C) 2010-2020 Adrian Perez <aperez@igalia.com>
 *
 * Distributed under terms of the MIT license.
 */

#define _POSIX_C_SOURCE 200809L

#include "deps/cflag/cflag.h"
#include "deps/clog/clog.h"
#include "deps/dbuf/dbuf.h"
#include "util.h"
#include <assert.h>
#include <errno.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <time.h>

#if !(defined(MULTICALL) && MULTICALL)
# define dlog_main main
#endif /* MULTICALL */

#ifndef TSTAMP_FMT
#define TSTAMP_FMT "%Y-%m-%d/%H:%M:%S"
#endif /* !TSTAMP_FMT */

#ifndef TSTAMP_LEN
#define TSTAMP_LEN (5 + 3 + 3 + 3 + 3 + 3)
#endif /* !TSTAMP_LEN */


static bool  timestamp  = false;
static bool  skip_empty = false;
static bool  buffered   = false;
static char *prefix     = NULL;
static int   log_fd     = -1;
static int   in_fd      = STDIN_FILENO;


static const struct cflag dlog_options[] = {
    CFLAG(string, "prefix", 'p', &prefix,
        "Insert the given prefix string between timestamps and logged text."),
    CFLAG(int, "input-fd", 'i', &in_fd,
        "File descriptor to read input from (default: stdin)."),
    CFLAG(bool, "buffered", 'b', &buffered,
        "Buffered operation, do not use flush to disk after each line."),
    CFLAG(bool, "timestamp", 't', &timestamp,
        "Prepend a timestamp in YYYY-MM-DD/HH:MM:SS format to each line."),
    CFLAG(bool, "skip-empty", 'e', &skip_empty,
        "Ignore empty lines with no characters."),
    CFLAG_HELP,
    CFLAG_END
};


static void
handle_signal (int signum)
{
    if (log_fd >= 0 && log_fd != STDOUT_FILENO && log_fd != STDERR_FILENO && !isatty(log_fd)) {
        if (safe_fsync(log_fd) == -1)
            clog_warning("Flushing log: %s", strerror(errno));
        if (safe_close(log_fd) == -1)
            clog_warning("Closing log: %s", strerror(errno));
        log_fd = -1;
    }

    /*
     * When handling HUP, we just sync and close the log file; in
     * other cases (INT, TERM) we have to exit gracefully, too.
     */
    if (signum != SIGHUP) {
        exit (EXIT_SUCCESS);
    }
}


int
dlog_main (int argc, char **argv)
{
    clog_init(NULL);

    struct dbuf overflow = DBUF_INIT;
    struct dbuf linebuf = DBUF_INIT;
    char *env_opts = NULL;
    struct sigaction sa;

    if ((env_opts = getenv ("DLOG_OPTIONS")) != NULL)
        replace_args_string (env_opts, &argc, &argv);

    const char *argv0 = cflag_apply(dlog_options, "[options] [logfile-path]", &argc, &argv);

    if (!argc) {
        log_fd = STDOUT_FILENO;
    } else if (safe_close(STDOUT_FILENO) == -1) {
		clog_warning("Closing stdout: %s", strerror(errno));
    }

    sigemptyset (&sa.sa_mask);
    sa.sa_flags = 0;

    sa.sa_handler = handle_signal;
    safe_sigaction ("HUP", SIGHUP, &sa);

    sa.sa_handler = handle_signal;
    safe_sigaction ("INT", SIGINT, &sa);

    sa.sa_handler = handle_signal;
    safe_sigaction ("TERM", SIGTERM, &sa);

    for (;;) {
        ssize_t bytes = freadline (in_fd, &linebuf, &overflow, 0);
        if (bytes == 0)
            break; /* EOF */

        if (bytes < 0)
            die ("%s: error reading input: %s\n", argv0, ERRSTR);

        if (!skip_empty || dbuf_size(&linebuf) > 1) {
            struct iovec iov[5];
            int n_iov = 0;

            char timebuf[TSTAMP_LEN+1];
            size_t timebuf_len;
            if (timestamp) {
                time_t now = time(NULL);
                struct tm *time_gm = gmtime (&now);

                if ((timebuf_len = strftime (timebuf, TSTAMP_LEN+1, TSTAMP_FMT, time_gm)) == 0)
                    die ("%s: cannot format timestamp: %s\n", argv0, ERRSTR);

                iov[n_iov++] = iov_from_data (timebuf, timebuf_len);
                iov[n_iov++] = iov_from_literal (" ");
            }

            if (prefix) {
                iov[n_iov++] = iov_from_string (prefix);
                iov[n_iov++] = iov_from_literal (" ");
            }

            iov[n_iov++] = iov_from_buffer (&linebuf);

            assert ((unsigned) n_iov <= (sizeof (iov) / sizeof (iov[0])));

            if (log_fd < 0) {
                if ((log_fd = safe_openatm(AT_FDCWD, argv[0], O_CREAT | O_APPEND | O_WRONLY, 0666)) < 0)
                    die ("%s: cannot open '%s': %s\n", argv0, argv[0], ERRSTR);
            }

            if (safe_writev(log_fd, iov, n_iov) < 0)
                clog_warning("Writing to log: %s", strerror(errno));

            if (!buffered && log_fd != STDOUT_FILENO && log_fd != STDERR_FILENO && !isatty(log_fd)) {
                if (safe_fsync(log_fd) != 0)
                    clog_warning("Flushing log: %s", strerror(errno));
            }
        }
        dbuf_clear(&linebuf);
    }

    if (log_fd >= 0 && safe_close(log_fd) == -1)
        clog_warning("Closing log: %s", strerror(errno));

    exit (EXIT_SUCCESS);
}

